Add gradle task to build platform specific distro

breskeby opened the pull request:
This adds the gradle task ':installLocalDistribution' that installs a distribution locally matching the local architecture.
 
Closes #53682

<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

breskeby committed:
Add gradle task to build platform specific distro

This adds the gradle task ':assemblePlatformArchive' that depends on the
platform specific distro archive tasks and allows using the same task for
building only the archive according to the local platform.

This delegates to

   :distribution:archives:windows-zip:assemble on windows
   :distribution:archives:linux-tar:assemble on linux
   :distribution:archives:darwin-tar:assemble on osx

Closes #53682

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

mark-vieira reviewed (commented):
I think we want to leverage our `DistributionDownloadPlugin` here. The plugin has all sort of smarts in it already to figure out how to resolve a requested distribution type to a project dependency or download. So I'd recommend we go that route so we don't have to duplicate the logic of which project to delegate to depending on the current os (that lives in the plugin already). Also, we'll probably want this task to extract the distro to some standard location. That way we can easily document that you can run this task, then look in `/extracted/dir` to find our distro. As it is, the build artifact is going to live in the build dir for the specific platform project, which is kind of confusing.

So let's do this:

1. Create a new script plugin for this and apply to the root project so we don't bloat the root build.gradle any further.
2. Apply the `elasticsearch.distribution-download` plugin.
3. Configure an `ElasticsearchDistribution` for the current platform, with `type = 'archive'` and `flavor = 'default'`.
4. Wire that distribution up to a copy task that copy the extracted distro to a known location `see `ElasticsearchDistribution.getExtracted()`.

breskeby commented:
Just to get a better understanding, why would we package and unpackage instead of just having a plain copy task that might share the same copy spec with the distro archive tasks? 

> On 30. Apr 2020, at 20:25, Mark Vieira <notifications@github.com> wrote:
> 
> ﻿
> @mark-vieira commented on this pull request.
> 
> I think we want to leverage our DistributionDownloadPlugin here. The plugin has all sort of smarts in it already to figure out how to resolve a requested distribution type to a project dependency or download. So I'd recommend we go that route so we don't have to duplicate the logic of which project to delegate to depending on the current os (that lives in the plugin already). Also, we'll probably want this task to extract the distro to some standard location. That way we can easily document that you can run this task, then look in /extracted/dir to find our distro. As it is, the build artifact is going to live in the build dir for the specific platform project, which is kind of confusing.
> 
> So let's do this:
> 
> Create a new script plugin for this and apply to the root project so we don't bloat the root build.gradle any further.
> Apply the elasticsearch.distribution-download plugin.
> Configure an ElasticsearchDistribution for the current platform, with type = 'archive' and flavor = 'default'.
> Wire that distribution up to a copy task that copy the extracted distro to a known location see ElasticsearchDistribution.getExtracted()`.
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

mark-vieira commented:
> Just to get a better understanding, why would we package and unpackage instead of just having a plain copy task that might share the same copy spec with the distro archive tasks?

That's a fantastic question 😉 .  Mainly because we just simply haven't optimized for that yet but it is very much worth doing as tar/untar adds considerable overhead. Right now we are basically just doing this all via artifact dependencies, so we grab the archive artifact, then unpack it. It'll need some work to make it more efficient.

breskeby committed:
Apply review feedback

- reuse distribution download plugin
- rename task to installLocalDistribution

breskeby commented:
@mark-vieira I applied your feedback

mark-vieira reviewed (approved):
LGTM 👍 

@rjernst want to contribute to some bike-shedding about naming here 😉

rjernst commented on gradle/local-distribution.gradle:
```diff
@@ -0,0 +1,41 @@
+  }
+}
+
+tasks.register('installLocalDistribution', Copy) {
```
perhaps we could make this easier to type as just `localDistro`? Should we also use `Sync` instead of `Copy`? Can't older files that were removed be left in place with `Copy`?

breskeby replied:
`localDistro` it is. And switched to use `Sync`

rjernst commented on gradle/local-distribution.gradle:
```diff
@@ -0,0 +1,41 @@
+tasks.register('installLocalDistribution', Copy) {
+  from(elasticsearch_distributions.local.extracted)
+  into("build/distribution/local")
+}
```
I think we should add a println that states the location of the expanded distribution

mark-vieira replied:
Good call, let's add a user friendly `doLast` which prints something like `Elasticsearch distribution installed to {location}`.

breskeby replied:
done.

pugnascotia commented on gradle/local-distribution.gradle:
```diff
@@ -0,0 +1,41 @@
+ */
+
+/**
+ * This script sets up a local distribution including.
```
Seems to be some words missing here.

breskeby replied:
fixed

breskeby committed:
Apply more feedback

- rename task name to localDistro
- use Sync instead of Copy task
- add user friendly output to point user to right directory

breskeby committed:
Update code comments

breskeby commented:
Example output https://gradle-enterprise.elastic.co/s/r7iukym2sivi4/console-log#L786

breskeby committed:
Change log level to lifecycle

rjernst reviewed (approved):
LGTM

mark-vieira commented:
@breskeby FYI, I added the `v7.9.0` label, corresponding to the `7.x` branch.
